src/halmos/sevm.py

```diff
@@ -1387,6 +1387,82 @@ def digest(self) -> bytes:
         return m.digest()
 
 
+class KeccakRegistry:
+    """A registry for tracking Keccak hash expressions and their corresponding values.
+
+    This class provides a dict-like interface for backward compatibility,
+    maintaining a mapping between hash values and their IDs (self._hashes).
+    This was the original functionality used for storing hash values and their corresponding IDs.
+
+    Additionally, this class provides an additional functionality for tracking
+    hash expressions and their generating values using an OffsetMap (self._exprs).
+    This allows for reverse lookups to find the original expression that
+    produced a given hash, even when the hash value is offset by some delta.
+    This is particularly useful for:
+    1. Local registry: Tracks hashes generated during execution
+    2. Precomputed registry: Contains known/common hash expressions
+
+    Example:
+        registry = KeccakRegistry()
+
+        # Storing hash values with ids (for backward compatibility)
+        registry[hash_value] = id_value
+        id_value = registry[hash_value]
+
+        # Tracking original hash expressions
+        registry.record(expr, hash_value)
+        original_expr = registry.reverse_lookup(hash_value)
+        # With offset
+        original_expr + delta = registry.reverse_lookup(hash_value + delta)
+    """
+
+    def __init__(self):
+        self._hashes: dict[BitVecRef, int] = {}  # hash -> id
+        self._exprs = OffsetMap()  # hash -> expr
+
+    def __getitem__(self, key: BitVecRef) -> int:
+        return self._hashes[key]
+
+    def __setitem__(self, key: BitVecRef, value: int) -> None:
+        self._hashes[key] = value
+
+    def __contains__(self, key: BitVecRef) -> bool:
+        return key in self._hashes
+
+    def __len__(self) -> int:
+        return len(self._hashes)
+
+    def items(self) -> Iterator[tuple[BitVecRef, int]]:
+        return self._hashes.items()
+
+    def keys(self) -> Iterator[BitVecRef]:
+        return self._hashes.keys()
+
+    def values(self) -> Iterator[int]:
+        return self._hashes.values()
+
+    def copy(self) -> "KeccakRegistry":
+        new_registry = KeccakRegistry()
+        new_registry._hashes = self._hashes.copy()
+        new_registry._exprs = self._exprs.copy()
+        return new_registry
```

Comment:
why would we need to copy this? seems to me this is an append-only structure so ideally copy(self) would return self

Reply:
um, no, we cannot share it across different paths, where there may exist different hash expressions that are evaluated to the same value.  we don't need deepcopy, but still need shallow copy.